### ISSUE TITLE
buffering toast disappears only after moving mouse

### ISSUE THREAD

**Issue by minusf** [iina/iina#4064]
**System and IINA version:**

- macOS 12.6.1
- IINA 1.3.1

**Expected behavior:**

buffering toast disappears

**Actual behavior:**

buffering toast stays visible, spinner spins but numbers are not being updated and the video is playing underneath:

<img width="168" alt="Screenshot 2022-11-11 at 16 22 07" src="https://user-images.githubusercontent.com/3632883/201371044-5bd59788-5ba0-4aa7-8c11-46904641f6cc.png">

**Steps to reproduce:**

open any youtube url

- [x] MPV does not have this problem.

**How often does this happen?**

in my limited testing it depended on how much/often buffering needed to happen.

**Comment by low-batt** [comment]
I previously tried to reproduce this and couldn't, so I didn't comment. I just encountered it while working on another issue. There is definitely a bug lurking where this notice does not get dismissed. This needs to be investigated.

**Comment by svobs** [comment]
It looks like `MainWindowController.hideUI()` might be causing the problem, because it calls `player.invalidateTimer()`, which kills the UI timer (the name `syncPlayTimeTimer` is no longer a good name because it does many things). The buffer indicator widget relies on `syncUITime()`, which for a network stream will call `mainWindow.updateNetworkState()`, which will update or hide it.

Maybe it can be fixed by adding another check to `hideUI()`, to make sure `bufferIndicatorView` is not still showing before killing the timer. But that opens up a new problem, which is how to decide when to kill the timer. And once the timer is killed, I don't see a mechanism for figuring out when to show `bufferIndicatorView` / start the timer again when there is a new buffering problem. The code appears to have been hitherto relying on that happening by virtue of the constant polling of the timer.

Perhaps the timer should just never be invalidated if the user is playing a network stream? They are, after all, presumably ok with using lots of network bandwidth while they are playing.

**Comment by low-batt** [comment]
On this:
> Perhaps the timer should just never be invalidated if the user is playing a network stream?

The killing of the timer has to do with IINA following the best practices in the [Energy Efficiency Guide for Mac Apps](https://developer.apple.com/library/archive/documentation/Performance/Conceptual/power_efficiency_guidelines_osx/BestPractices.html#//apple_ref/doc/uid/TP40013929-CH38-SW1). In this case:
> Ensure your app is absolutely idle when not in use

Various changes have been made in an attempt to stop IINA from running operations that aren't needed. Timers are expensive, so the code is trying to not run this timer when it is not needed. If you pause playback while streaming you should expect to see that IINA is hardly using any CPU. So always running the timer while streaming is not desirable.

Seems like the code is failing to take into account the need to manage the buffer indicator when streaming. Likely IINA will need to run a timer while streaming, but it should be shut down when not needed, such as when paused.

And, ideally IINA should be smart about this. Changing the precision of the times shown in the OSC causes that timer run on a `0.04` interval. I would argue that such a short cycle is not needed to manage the buffer indicator. When the OSC is hidden and the timer is only needed to manage the buffer indicator the timer should be adjusted to use a longer interval.

I think this one needs a non-trivial "smart" fix.

Did you want to take this one?

**Comment by svobs** [comment]
> Seems like the code is failing to take into account the need to manage the buffer indicator when streaming. Likely IINA will need to run a timer while streaming, but it should be shut down when not needed, such as when paused.

Oh right..._pause_... Had some drinks earlier last night but guess was still past the [Ballmer peak](https://websitehurdles.com/do-programmers-drink-alcohol/)...

> Did you want to take this one?

Alright...yeah, I can give it a shot.

**Comment by svobs** [comment]
> Various changes have been made in an attempt to stop IINA from running operations that aren't needed. Timers are expensive, so the code is trying to not run this timer when it is not needed. If you pause playback while streaming you should expect to see that IINA is hardly using any CPU.

> And, ideally IINA should be smart about this. Changing the precision of the times shown in the OSC causes that timer run on a 0.04 interval. I would argue that such a short cycle is not needed to manage the buffer indicator. When the OSC is hidden and the timer is only needed to manage the buffer indicator the timer should be adjusted to use a longer interval.

I have been running experiments with the Time Profiler, and as it always seems with profiling, the biggest offenders are not quite what we assumed. One disclaimer: by their very nature of living between threads, it's not that easy to quantify the performance of the Timer mechanisms. Although you can see a lot of the system calls in the reports, it's not super clear how to interpret them because there's no visibility into Apple's code. To make things worse, the Time Profiler numbers are not comprehensive but are the result of periodic sampling at about 1000 times/second (supposedly), so large numbers of small calls may be underrepresented. But by playing with various scenarios and looking at the total time used and how it changed, I was able to gain some insights. Hopefully I will get around to organizing the data and presenting a report while it's still fresh in my head. But for now, a couple major takeaways:
1. The `Timer` mechanisms, in sum, don't appear to be as big of a performance impact as uses of `DispatchQueue.main.async()`, which are still significant even when the player is paused. It would be good to try to avoid adding too many more, and maybe try to consolidate existing ones where possible.
2. When the video is playing and the OSC is still showing (thus still syncing the UI), by far the most expensive operation (other than the video itself) appears to be `PlaySlider.drawRect()`, which takes up over half the CPU usage on the main thread, and this increases linearly with the frequency of the timer. By contrast, redrawing the elapsed/duration text consumes 1/10th of that. So it would be smarter to first attack the play slider, maybe by adding some code which only does some quick basic geometric calculations and _only_ redraws it if its width going to change by more than 0.5 pixels or so.

Will leave those for some future issue. Going back to this issue - I came up with a solution I'm happy with and will post a PR shortly.

**Comment by low-batt** [comment]
My worries about the energy use of timers comes from Apple's Energy Efficiency Guide for Mac Apps. One of Apple's recommendations is to [Minimize Timer Usage](https://developer.apple.com/library/archive/documentation/Performance/Conceptual/power_efficiency_guidelines_osx/Timers.html#//apple_ref/doc/uid/TP40013929-CH5-SW1). Apple also discusses the need to make sure your app is completely idle when it is not performing work.

Mostly I've been trying to make sure IINA is not drawing energy when it should be totally idle, as in when playback is paused. Of course the big consumer is playback. 

Of course energy use is still a side issue. New defects have been reported since the upgrade of dependencies.

I will take a look at the PR when I get a chance.

**Comment by low-batt** [comment]
This was fixed in IINA 1.4.0-beta1.
